Framework/API/src/InfoComponentVisitor.cpp

```diff
@@ -4,74 +4,113 @@
-  bank.getChildren(assemblyChildren, false /*is recursive*/);
+  assembly.getChildren(assemblyChildren, false /*is recursive*/);
 
+  const size_t detectorStart = m_detectorIndices.size();
   for (const auto &child : assemblyChildren) {
     // register everything under this assembly
-    child->registerContents(*this, localDetectorIndexes);
+    child->registerContents(*this);
   }
+  const size_t detectorStop = m_detectorIndices.size();
 
-  /*
-   * The following allows us to refer to ALL nested
-   * detectors from any component in the tree. Otherwise these could
-   * be hidden by sub-tree assemblies.
-   */
-  parentDetectorIndexes.reserve(parentDetectorIndexes.size() +
-                                localDetectorIndexes.size());
-  parentDetectorIndexes.insert(parentDetectorIndexes.end(),
-                               localDetectorIndexes.begin(),
-                               localDetectorIndexes.end());
-  m_componentDetectorIndexes.emplace_back(localDetectorIndexes);
-  m_componentIds.emplace_back(bank.getComponentID());
+  m_ranges.emplace_back(std::make_pair(detectorStart, detectorStop));
+
+  // For any non-detector we extend the m_componetIds from the back
+  m_componentIds.emplace_back(assembly.getComponentID());
 }
 
-void InfoComponentVisitor::registerGenericComponent(const IComponent &component,
-                                                    std::vector<size_t> &) {
+/**
+ * @brief InfoComponentVisitor::registerGenericComponent
+ * @param component : IComponent being visited
+ */
+void InfoComponentVisitor::registerGenericComponent(
+    const IComponent &component) {
   /*
    * For a generic leaf component we extend the component ids list, but
    * the detector indexes entries will of course be empty
    */
-  m_componentDetectorIndexes.emplace_back(std::vector<size_t>());
+  m_ranges.emplace_back(std::make_pair(0, 0)); // Represents an empty range
   m_componentIds.emplace_back(component.getComponentID());
 }
-void InfoComponentVisitor::registerDetector(
-    const IDetector &detector, std::vector<size_t> &parentDetectorIndexes) {
 
-  /*
-   * Add the detector index to the parent, but also report the "self" detector
-   * index
-   * if te component index passed is one of a detector.
-   */
-  auto detectorIndex = m_detectorInfo.indexOf(detector.getID());
-  parentDetectorIndexes.push_back(detectorIndex);
-  m_componentDetectorIndexes.emplace_back(
-      std::vector<size_t>(1, detectorIndex));
-  m_componentIds.emplace_back(detector.getComponentID());
+/**
+ * @brief InfoComponentVisitor::registerDetector
+ * @param detector : IDetector being visited
+ */
+void InfoComponentVisitor::registerDetector(const IDetector &detector) {
+
+  const auto detectorIndex =
+      m_detectorIdToIndexMapperFunction(detector.getID());
+
+  /* Already allocated we just need to index into the inital front-detector
+   * part of the collection
+  */
+  m_componentIds[m_detectorCounter++] = detector.getComponentID();
```

Comment:
I suppose this does not guarantee that the order of detector indices is the same as the order of component indices? That is detector ID 17 might be detector index 4, but component index 8?